security/fma-holocene.md

```diff
@@ -0,0 +1,196 @@
+- (Smart Contracts) Update to the MIPS contract
+
+Each change has it's own section below with a list of Failure Modes.
+
+Below are references for this project:
+
+- [PID: Holocene hardfork upgrade](https://www.notion.so/PID-Holocene-hardfork-upgrade-00ee1ffc414a407088fdb49841771527?pvs=21)
+- [Github tracker](https://github.com/orgs/ethereum-optimism/projects/84/views/6)
+- [Specs](https://specs.optimism.io/protocol/holocene/derivation.html?highlight=holocene#holocene-derivation)
+
+## Holocene Derivation
+
+### The batcher violates the new (stricter) ordering rules
+
+#### Description
+
+If the batcher:
+
+- sends transactions out of order, or
+- orders frames within transactions out of order, or
+- orders batches within a channel out of order
+
+This will lead to the consensus client hitting error conditions when it loads the frames from L1. According to the spec, frames will be dropped in order to maintain a contiguous frame queue, and out-of-order batches will be dropped at a later stage in the derivation pipeline. The safe chain would halt until the next batch can be resubmitted in order.
+
+#### Risk Assessment
+
+low severity / low likelihood
+
+#### Mitigations
+
+The batcher implementation could:
+
+- leverage nonce management to avoid sending transactions out of order in the first place (see [spec](https://specs.optimism.io/protocol/holocene/derivation.html?highlight=holocene#batcher-hardening))
+- be hardened so detect the chain halt and resubmit the frames which were dropped (see [spec](https://specs.optimism.io/protocol/holocene/derivation.html?highlight=holocene#batcher-hardening)) either:
+  - as a part of normal operation
+  - as a part of its startup behavior (i.e. to be triggered by a restart)
+  - as a part of an emergency mode triggered by an admin API, this could allow for manual intervention
+- continually check for contiguity in its internal state, and panic or reset if this is violated (possibly then triggering some recovery behavior)
+
+#### Detection
+
+L2 safe chain halt
+
+#### Recovery Path(s)
+
+We would need to get the batcher to resubmit the transaction sent out of order. No governance or hardfork needed to recover. A simple batcher restart would cause the batcher to re-batch the unsafe chain, which should lead to recovery. Moreover, we could also temporarily operate the batcher with reduced tx sending concurrency, which should avoid out of order txs.
+
+### Derivation Deadlock (either specification or op-node bug)
+
+#### Description
+
+It is possible that either i) the Holocene Derivation Specification or ii) the consensus client implementation has overlooked a corner case such that derivation simply halts when that corner case arises.
+
+#### Risk Assessment
+
+**high severity** / medium likelihood
+
+#### Mitigations
+
+Multi client testing should help surface any consensus client bugs, possibly including fuzzing of some description.
```

Comment:
Comment from @clabby: important we take time to create action tests for this, and ideally implement chaos or fuzz testing.

Reply:
The action tests will be run with Kona as well as the op-node derivation. 

Let's add a concrete action item: When we deploy to the devnet, we will want to have geth and reth running. 

We will also be moving towards kurtosis based fuzzing for multiple clients. This will be a nice to have action item, we should add it to the generic FMA doc and make sure we don't do too many more hardforks without having this in place.

Reply:
[2f5dccc](https://github.com/ethereum-optimism/design-docs/pull/127/commits/2f5dcccf9868012e3c72261de80f2102177b2243)